cmd/snap/cmd_components.go

```diff
@@ -0,0 +1,135 @@
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package main
+
+import (
+	"fmt"
+	"sort"
+
+	"github.com/jessevdk/go-flags"
+
+	"github.com/snapcore/snapd/client"
+	"github.com/snapcore/snapd/i18n"
+	"github.com/snapcore/snapd/snap/naming"
+)
+
+var shortComponentsHelp = i18n.G("List available and installed components for installed snaps")
+var longComponentsHelp = i18n.G(`
+The components command displays a summary of the components that are installed
+and available for the set of currently installed snaps.
+
+Components for specific installed snaps can be queried by providing snap names
+as positional arguments.
+`)
+
+type cmdComponents struct {
+	clientMixin
+	Positional struct {
+		Snaps []installedSnapName `positional-arg-name:"<snap>"`
+	} `positional-args:"yes"`
+}
+
+func init() {
+	addCommand("components",
+		shortComponentsHelp,
+		longComponentsHelp,
+		func() flags.Commander { return &cmdComponents{} },
+		nil,
+		[]argDesc{{
+			name: i18n.G("<snap>"),
+			desc: i18n.G("Snaps to consider when listing available and installed components."),
+		}},
+	)
+}
+
+func (x *cmdComponents) Execute(args []string) error {
+	if len(args) > 0 {
+		return ErrExtraArgs
+	}
+
+	names := installedSnapNames(x.Positional.Snaps)
+	snaps, err := x.client.List(names, nil)
+	if err != nil {
+		if err == client.ErrNoSnapsInstalled {
+			if len(names) == 0 {
+				fmt.Fprintln(Stderr, i18n.G("No snaps are installed yet. Try 'snap install hello-world'."))
+				return nil
+			} else {
+				return ErrNoMatchingSnaps
+			}
+		}
+		return err
+	}
+
+	anyComps := false
+	for _, snap := range snaps {
+		anyComps = anyComps || len(snap.Components) > 0
+	}
+
+	if !anyComps {
+		fmt.Fprintln(Stderr, i18n.G("No components are available for any installed snaps."))
+		return nil
+	}
+
+	sort.Sort(snapsByName(snaps))
+
+	w := tabWriter()
+	fmt.Fprintln(w, i18n.G("Component\tStatus\tType"))
+	for _, snap := range snaps {
+		for _, comp := range sortedComponents(snap.Components) {
+			// note that snap.Name is actually an instance name, and this isn't
+			// how we'd usually use a naming.ComponentRef. however, presenting
+			// users with a string that they can copy-paste into a "snap
+			// install" command seems useful
+			name := naming.NewComponentRef(snap.Name, comp.Name).String()
+			status := "available"
+			if comp.InstallDate != nil {
+				status = "installed"
+			}
+			fmt.Fprintf(w, "%s\t%s\t%s\n", name, status, comp.Type)
+		}
+	}
+	w.Flush()
+
+	return nil
+}
+
+func sortedComponents(comps []client.Component) []client.Component {
+	copied := make([]client.Component, len(comps))
+	copy(copied, comps)
+
+	// installed components are put first, followed by available components.
+	// components within those groups are sorted lexicographically
+	sort.Slice(copied, func(i, j int) bool {
```

Comment:
in some other places doing the sorting we often used `sort.Sort` with an interface named to indicate what kind of sorting it does. I suppose in this case you could to:

```
type componentsByInstallStatusAndSnapName []client.Component

type (c componentsByInstallStatusAndSnapName) Len() int { return len(c) }
type (c componentsByInstallStatusAndSnapName) Swap(i, j int) { c[i], c[j] = c[j], c[i] }
type (c componentsByInstallStatusAndSnapName) Less(i, j int) bool {
	left, right := c[i], c[j]

	if left.InstallDate == nil && right.InstallDate != nil {
		return false
	}

	if left.InstallDate != nil && right.InstallDate == nil {
		return true
	}

	return left.Name < right.Name
}
```

Reply:
These result in the same thing, however using sort.Slice is more concise, I think? Generally, I've seen sort.Sort used for custom containers (or from before sort.Slice became a thing). I can use a named function to create the closure if the concern is readability.